src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -298,9 +298,16 @@ internal abstract class UnboundLambdaState
     {
         private UnboundLambda _unboundLambda; // we would prefer this readonly, but we have an initialization cycle.
         protected readonly Binder binder;
-        private readonly ConcurrentDictionary<NamedTypeSymbol, BoundLambda> _bindingCache = new ConcurrentDictionary<NamedTypeSymbol, BoundLambda>();
 
-        private readonly ConcurrentDictionary<ReturnInferenceCacheKey, BoundLambda> _returnInferenceCache = new ConcurrentDictionary<ReturnInferenceCacheKey, BoundLambda>();
+        [PerformanceSensitive(
+            "https://github.com/dotnet/roslyn/issues/23582",
+            Constraint = "Avoid " + nameof(ConcurrentDictionary<NamedTypeSymbol, BoundLambda>) + " which has a large default size, but this cache is normally small.")]
+        private ImmutableDictionary<NamedTypeSymbol, BoundLambda> _bindingCache = ImmutableDictionary<NamedTypeSymbol, BoundLambda>.Empty;
+
+        [PerformanceSensitive(
+            "https://github.com/dotnet/roslyn/issues/23582",
+            Constraint = "Avoid " + nameof(ConcurrentDictionary<NamedTypeSymbol, BoundLambda>) + " which has a large default size, but this cache is normally small.")]
+        private ImmutableDictionary<NamedTypeSymbol, BoundLambda> _returnInferenceCache = ImmutableDictionary<NamedTypeSymbol, BoundLambda>.Empty;
```

Comment:
>NamedTypeSymbol [](start = 110, length = 15)

The change in the key type looks wrong.

Reply:
I wouldn't call it so much wrong as replacing one apparently unimportant optimization (sharing the cache for semantically identical delegate types) for another apparently important optimization. I believe the transformation preserves correctness of the compiler. Do you, @AlekseyTs, believe that change is semantically incorrect?

Reply:
This one is going to be important to understand. According to the previous implementation of `BindForReturnTypeInference`, returning a new result from `ReallyInferReturnType` for the same delegate type is acceptable in the case of a cache miss (note the use of `TryAdd` instead of `GetOrAdd`). I verified that the conversion from `NamedTypeSymbol` to `ReturnInferenceCacheKey` is a deterministic function, but incorrectly treated it as one-to-one.

Reply:
Neither description of the change, nor comments mentioned the type change for the key. The motivation behind the change was not explained, no numbers supporting the change were provided. I am talking exclusively about the type change for the key. If you have numbers supporting that the type change **on its own** leads to noticeable improvement, and you are confident that the change is not going to regress other scenarios, then we should get rid of the ReturnInferenceCacheKey type altogether.